### ISSUE TITLE
Issue linking React as a target dependency in Xcode for React Native iOS apps: 'RCTLog.h' file not found

### ISSUE THREAD

**Issue by farhankassam2** [react-native-community/discussions-and-proposals#287]
Dear React Native community,

I write to you this message to discuss with you a major issue I have been struggling with for the past two weeks related to creating an archive for our project's React Native iOS application.

Before I go in-depth with the issue, I want to mention that our team lead is enrolled in the Apple Developer Program for which we have a valid distribution signing certificate and distribution provisioning profiles associated with our application: this is necessary in archiving an iOS app for distribution.

I also want to state that I have already followed the steps in this [Prepare for app distribution page](https://help.apple.com/xcode/mac/current/#/dev91fe7130a)  to prepare our app for distribution and all the requirements have been met to create an archive for our app. I have also looked around the [Developer Documentation](https://developer.apple.com/documentation) and "Apple Developer Forums" for a solution to this and tried precisely twelve different solutions to fix this depending on posts from other users of the forum as well as from Github issues and Stackoverflow links, but none seemed to resolve this.

After contacting the Apple Developer Technical support through our Apple Developer Program paid account (we submitted a TSI (technical support request), they stated that Apple "doesn’t support third-party libraries or dependency tools". After, they suggested me to contact "the React Native development community to identify the cause of and a fix for this issue". Hence, I decided to make a post on here to see if this is a major issue and perhaps if the way in which the `react-native` or `react` npm packages are linked to Xcode in RN iOS projects needs to be changed or updated, considering the many people who have encountered this issue in the past: this is evident from multiple posts I have seen as linked Github issues, on Stackoverflow and the Apple Developer Forum.

The technical issue is as follows: it seems to be appearing in the `RNBranchConfig.m` file within the react-native-branch package for iOS specifically in lines 9 and 23 that `'React/RCTLog.h' file not found` and `Use of undeclared identifier 'RNBRANCH_VERSION'` respectively in the links below:

https://github.com/BranchMetrics/react-native-branch-deep-linking-attribution/blob/3b3b3df9b72bc832d1c525627253ec9ac7955c1d/ios/RNBranchConfig.m#L9

https://github.com/BranchMetrics/react-native-branch-deep-linking-attribution/blob/3b3b3df9b72bc832d1c525627253ec9ac7955c1d/ios/RNBranchConfig.m#L23

Now, you may be wondering that this issue may be related to the package provider but for the former error, it seems as though the package cannot detect the React target dependencies and in particular, this `RCTLog.h` file which is located within the React Native framework. Thus, it seems to me that this may be an issue within React Native itself, somehow not able to link to the Xcode project.

A screenshot of the error message is provided below:
<img width="1028" alt="Screen Shot 2020-10-14 at 2 18 48 PM" src="https://user-images.githubusercontent.com/22672953/96046525-7b986e00-0e28-11eb-8df7-700e7a5c5c95.png">

To reproduce the issue, do the following:
1) Open <project>.xcodeproj in Xcode and add the `RNBranch.xcodeproj` manually into the Library directory as follows:
<img width="445" alt="Screen Shot 2020-10-14 at 3 30 05 PM" src="https://user-images.githubusercontent.com/22672953/96052369-4133ce80-0e32-11eb-80a7-dd7cc47e8abc.png">

I did the above to remove the `-lBranch not found` error I was experiencing in Xcode before as below:
![Screen Shot 2020-10-11 at 2 34 52 PM copy](https://user-images.githubusercontent.com/22672953/96052886-3e85a900-0e33-11eb-9dd0-c5cc31ebbe68.png)
2) Go to `Product > Clean Build Folder` to start the build process afresh.

3) Make sure that the device to run the target app on is `Generic iOS Device` in the top-left corner next to the Run symbol as shown below:
<img width="485" alt="Screen Shot 2020-10-14 at 3 15 32 PM" src="https://user-images.githubusercontent.com/22672953/96051353-2b250e80-0e30-11eb-8767-e336b3dce6ae.png">

4) Click on `Product > Archive` to start the archiving process of generating the `.ipa` file which compresses the app into a file that can be distributed to other software (I am using this so that I can distribute the app to BETA testers). This is shown below:
<img width="575" alt="Screen Shot 2020-10-14 at 3 17 13 PM" src="https://user-images.githubusercontent.com/22672953/96051479-73443100-0e30-11eb-951d-589647868328.png">

5) For me, during the archiving process, it pops up a terminal window showing that dependency graphs have been loaded, as it should as follows:
<img width="592" alt="Screen Shot 2020-10-14 at 3 18 59 PM" src="https://user-images.githubusercontent.com/22672953/96051556-a25aa280-0e30-11eb-9cab-c09ff23a635a.png">

After a minute or so, it shows that there are two compile-time errors, both in the same aforementioned file. I will show the full image here after the failed archiving process showing all the details:
<img width="1380" alt="Screen Shot 2020-10-14 at 3 22 39 PM" src="https://user-images.githubusercontent.com/22672953/96051839-3462ab00-0e31-11eb-8d3f-b75d46c91c62.png">

These are the solutions I have tried so far:
1) Adding `React` target dependency to to my project's target as below:
<img width="1013" alt="Screen Shot 2020-10-14 at 2 21 52 PM" src="https://user-images.githubusercontent.com/22672953/96046705-bdc1af80-0e28-11eb-9b27-5c4631db1490.png">

2) Adding React as a `target` itself to the project schemes and unchecking parallelize build as below:
<img width="901" alt="Screen Shot 2020-10-14 at 2 36 40 PM" src="https://user-images.githubusercontent.com/22672953/96047956-bf8c7280-0e2a-11eb-97c8-cd00b47353af.png">

3) Manually adding the `React.xcodeproj` as a Library to use in the project under `<project>/Library` and then adding `libReact.a` into the following:
a) `Build Phases > Link Binary` with Libraries as below:
<img width="1009" alt="Screen Shot 2020-10-14 at 2 40 14 PM" src="https://user-images.githubusercontent.com/22672953/96048269-3d507e00-0e2b-11eb-893a-c0012baa542e.png">
b) `General > Frameworks, Libraries and Embedded Content` as below:
<img width="1010" alt="Screen Shot 2020-10-14 at 2 41 57 PM" src="https://user-images.githubusercontent.com/22672953/96048446-7b4da200-0e2b-11eb-8eb0-4fcdf8e209bb.png">

The `React.xcodeproj` was added in the following `Library` directory manually:
<img width="526" alt="Screen Shot 2020-10-14 at 2 43 32 PM" src="https://user-images.githubusercontent.com/22672953/96048674-c667b500-0e2b-11eb-8667-72288510e489.png">

4) Commenting out line 9 and line 23 in the first image to see if those two are even needed. It turns out that even after this, the archiving fails with the other remaining warnings. 

5) Trying to use the following `react-native` package versions: `0.60.0`, `0.61.2`, `0.62.1` and `0.63.3` with `react` peer dependency versions: `16.13.1` and `16.90` and in different combinations with each other to see if perhaps the file can be found in different versions of React. In some Github posts I read that the issue occurs only with RN versions lower than 0.61, which is why I chose to also downgrade the version. Initially, I had been using `0.62.1` and after that I was playing around with the various versions aforementioned.

6) Making sure that `pod React` is correctly added in my Podfile to link with the correct podspec as below is a line in my Podfile:
<img width="836" alt="Screen Shot 2020-10-14 at 2 50 20 PM" src="https://user-images.githubusercontent.com/22672953/96049270-a389d080-0e2c-11eb-8486-6f051d348e5d.png">

7) Adding `subspecs` for the various other `React` dependencies to my podfile and then running `pod install` within `ios` directory to ensure that the subspecs are also included as below:
<img width="983" alt="Screen Shot 2020-10-14 at 2 53 20 PM" src="https://user-images.githubusercontent.com/22672953/96049545-08452b00-0e2d-11eb-8aba-7b90247f3d2d.png">

When I tried the above, it threw me an error after doing `pod install` that some of the podspec files like the ones corresponding to `RCTLinkingIOS` and `DevSupport` cannot be found and therefore, `pod install` would fail:

8) I ensured I have `$(inherited)` in the header search paths and I also added `${SRCROOT}/../../../ios/Pods/Headers` to the header search paths so that the project can detect the React dependency and associated files. I tried both of these with `recursive` and `non-recursive` strategies just for the sake of it. This is show below:
<img width="908" alt="Screen Shot 2020-10-14 at 3 01 22 PM" src="https://user-images.githubusercontent.com/22672953/96050358-53137280-0e2e-11eb-96f3-4baaa573ce4f.png">

9) I have also ensured that I have `$(inherited)` as `non-recursive` in the Library search paths for both debug and release as shown below:
<img width="1003" alt="Screen Shot 2020-10-14 at 3 05 44 PM" src="https://user-images.githubusercontent.com/22672953/96050674-e0ef5d80-0e2e-11eb-975c-da50db7fbc28.png">

10)I have tried setting the `Always Search User Paths (Deprecated)` to both `Yes` and `No`. I tried `Yes` because according to Xcode help it states that enabling it may allow Xcode to search for libraries in the format `<React/RCTLog.h>` as well as in the format `"RCTLog.h"` as stated in the Xcode help window below (I am attaching it here for reference):
<img width="395" alt="Screen Shot 2020-10-14 at 3 09 28 PM" src="https://user-images.githubusercontent.com/22672953/96050973-6ffc7580-0e2f-11eb-8de1-26f6abab3c7d.png">

The setting is located under `Build Settings > Search Paths` as show below:
<img width="855" alt="Screen Shot 2020-10-14 at 3 07 30 PM" src="https://user-images.githubusercontent.com/22672953/96051016-81458200-0e2f-11eb-9ef1-14c1a8951d21.png">

11) I downloaded the `RCTLog.h` file provided by React in this [LINK](https://github.com/facebook/react-native/blob/master/React/Base/RCTLog.h) and manually added it into the `RNBranch.xcodeproj` directory too as show below:
<img width="436" alt="Screen Shot 2020-10-14 at 4 12 04 PM" src="https://user-images.githubusercontent.com/22672953/96055107-13ea1f00-0e38-11eb-9113-1be6df7a947e.png">

I then changed the import to: `#import "RCTLog.h"` in line 9 and line 23 to: `NSString * const RNBNC_PLUGIN_VERSION = @"5.0.0";` of `RNBranchConfig.m` to let the file know that the `RCTLog.h` is located within its own Xcode project and that the current react native branch version is `5.0.0` (hardcoded) as show below:
<img width="951" alt="Screen Shot 2020-10-14 at 4 14 30 PM" src="https://user-images.githubusercontent.com/22672953/96055268-704d3e80-0e38-11eb-95c5-19e9c81beaa2.png">

After I try to archive the app; it fails again.

12) The last solution I tried was following the steps outlined on this Github issue post, but instead of linking `react-native-branch`, I tried linking and then doing `pod install` for both `react@16.14.0` and `react-native@0.63.3`, hoping that when I run `react-native link react-native` and `react-native link react`, the machine will think that cocoapods is not installed, which apparently according to the link, is able to correctly link the React dependencies into the Xcode project and detected by packages using `React.xcodeproj`. For further details on the solution I tried, visit this issue comment: [LINK](https://github.com/BranchMetrics/react-native-branch-deep-linking-attribution/issues/361#issuecomment-415149270). I followed the same steps even when linking the `react-native-branch` package, as that is the package causing this issue. But after much trial and error, the same error pops when I go to archive.

Now that I have described the all the solutions I have tried, you can see that I have tried so many different things to get `RNBranchConfig.m` to detect `RCTLog.h` in the React dependency, but NONE worked! I am currently out of options and that is why I ask for some help. We need to get our app release for BETA testing within two weeks now so I also have a time constraint, which, as a developer, is quite pressurizing I hope you can understand. I hope the issues in this `react-native-branch` can be resolved sooner than later to make it a reliable long-term package to use for deep linking in iOS.

Below, find details about the version of the various tools, software, packages and libraries I am using while I encounter this error so that you may assist me faster:
- `react-native`: `^0.60.0`, `^0.61.2`, `^0.62.1` and `^0.63.3`
- `react`: `^16.13.1`,`^16.90` and `^16.14.0`
- `react-native-branch`: `^5.0.0`
- `cocoapods`: `cocoapods-1.8.4`
- `node`: `v12.14.1`
- `npm`: `6.13.4`
- `Xcode`: `11.4`
- `macOS`: `Catalina 10.15.5`
- `ruby`: `ruby-2.6.3`
- `react-native-cli`: `2.0.1`

I also have an Xcode archive log, which I shall attach below as a text file: 
[Archive PhoneMe_2020-10-14T23-10-15.txt](https://github.com/BranchMetrics/react-native-branch-deep-linking-attribution/files/5383493/Archive.PhoneMe_2020-10-14T23-10-15.txt)

I also have a build log exported from the IDE when trying to build the React Native iOS app from the IDE here: [build_log.pdf](https://github.com/BranchMetrics/react-native-branch-deep-linking-attribution/files/5383509/build_log.pdf).

Any help is greatly appreciated.

**Comment by kelset** [comment]
👋 there - as stated in the README and the issue template, this repo is for long-form discussions. Since your is a technical problem, should be posting this issue on the main [react native repo](https://github.com/facebook/react-native).

I appreciate the details that you are providing but the general vibe I get from your report is that something in the configuration of your precise project is not well-linked. And by that in particular I mean it's kind of weird that you were able to "quickly" test so many different version of react-native for your project - are you sure you have always followed all the [upgrade-helper](https://react-native-community.github.io/upgrade-helper/) steps when moving between versions? Also, since version 0.60 you are [not supposed to run](https://reactnative.dev/blog/2019/07/03/version-60#native-modules-are-now-autolinked) `react-native link` anymore, which to me is another flag that something within your project might be misconfigured.

I would highly recommend that you try to reproduce your issue in a freshly init-ed project and post that along your issue in the main repo.

**Comment by radko93** [comment]
Personal tip: at this point, I would suggest you to create a new RN project and try to recreate your setup (but do not link anything manually! I don't know any modules that you need to link manually anymore, don't use 'react-native link` to link native libraries). It might be easier than upgrading trough multiple versions of React Native.

**Comment by farhankassam2** [comment]
> Also, since version 0.60 you are [not supposed to run](https://reactnative.dev/blog/2019/07/03/version-60#native-modules-are-now-autolinked) `react-native link` anymore, which to me is another flag that something within your project might be misconfigured.

Thank you for your reply @kelset . I understand I will make a post on the react native repo directly than. As for the question you asked, I had initially also tried building the project without `react-native link`. I only tried that solution in item number 12) in the list of solutions I explored, as one Github issue post directed me to do so. I wanted to try it out. Otherwise in the remaining of the 11 solutions I tried, I did not run `react-native link` to link the `react-native-branch` package or even to link `react-native` or `react` itself. 



> are you sure you have always followed all the [upgrade-helper](https://react-native-community.github.io/upgrade-helper/) steps when moving between versions?

I did not follow the steps highlighted in the link you sent me. However, even without upgrading any versions in `react-native: ^0.62.1`, I still experienced the same issue. Therefore, I do not think the problem is associated with changing the `react-native` versions when this issue persisted even before I tried various other versions.

 

> I would highly recommend that you try to reproduce your issue in a freshly init-ed project and post that along your issue in the main repo.

I will go ahead with your suggestion and try creating a fresh new RN project and adding the `react-native-branch` package and see if I still get the error. I will also make an issue post in the react native repo: thank you for the advise.

**Comment by farhankassam2** [comment]
> I will go ahead with your suggestion and try creating a fresh new RN project and adding the `react-native-branch` package and see if I still get the error. I will also make an issue post in the react native repo: thank you for the advise.

I followed your advice and managed to reproduce the error with minimal code. It seems that the Branch package itself is not linking correctly and thus, detecting the `RCTLog.h` file. Since we cannot share our project's repository publicly, I created a Github repo out of this new fresh RN project, which can be used to reproduce the error on anyone's local machine. Here it is: https://github.com/farhankassam2/PNM_179_Test_Project/.

I have also posted on the react native repo here: https://github.com/facebook/react-native/issues/30221. Thanks.
